ggml-quants.c

```diff
@@ -3306,6 +3324,53 @@ size_t quantize_q8_0(const float * restrict src, void * restrict dst, int64_t nr
     return nrow * row_size;
 }
 
+size_t quantize_i2_s(const float * restrict src, void * restrict dst, int64_t nrow, int64_t n_per_row, const float * quant_weights) {
+    // 2 bits per weight
+    UNUSED(quant_weights);
+
+    size_t row_size = ggml_row_size(GGML_TYPE_I2_S, n_per_row);
+
+    int n = nrow * n_per_row;
+
+    // f32 -> q8
+    double i2_scale = 0;
+    for (int i=0; i<n; i++) {
+        if (fabs((double)(src[i])) > 1e-6) {
+            i2_scale = (double)src[i];
```

Comment:
Thanks for remind!
Actually, at this time, the weight matrix only contains three different value : (1, 0, -1) * scale (for example, 0.66, 0, -0.66). So, as long as we don't pick 0 as scale, any non-zero value could be the scale. If we pick 0.66 as scale, we will transform (0.66, 0, -0.66) to (1, 0, -1), if we pick -0.66 as scale, we will transform (0.66, 0, -0.66) to (-1, 0, 1).
I can add a break to the loop to pick the first non-zero value to avoid useless looping.

Reply:
> I can add a break to the loop to pick the first non-zero value to avoid useless looping.

This would slightly complicate the eventual Numpy implementation in `gguf-py/gguf/quants.py` for `convert-hf-to-gguf.py`. Quantization doesn't need to be particularly fast, but it needs to be reproducible. If all values are the same, picking the absmax might be fine then.

(It's dequantization that needs to be fast for good inference speed)

Reply:
I got it. I will change it into absmax to make it more reproducible.

Reply:
@Eddie-Wang1120, actually, in section 2 of the BitNet 1.58b paper, they specifically say they use absmean:

> To constrain the weights to -1, 0, or +1, we adopt an *absmean* quantization function. It first scales the weight matrix by its average absolute value, and then round each value to the nearest integer among {-1, 0, +1}

See <https://arxiv.org/html/2402.17764v1#S2>

Reply:
And an interesting fact is, if we don't pre-quantize weights to {-1, 0, +1} * scale, the tokens generated will be wrong. That's why I put the absmean quantization (weight pre-quantization) in convert-hf-to-gguf.py, otherwise we'll get a meaningless fp32/fp16 gguf model.